### PR TITLE
add trainer classes and rmsprop

### PR DESCRIPTION
**Pull request by eiofmv**

### CODE REVIEW

File: get_csv.sh [hunk 1]
```diff
@@ -1,2 +1,5 @@
-wget https://cernbox.cern.ch/index.php/s/SAHcpBDTp4P9jfO/download
-tar -xzf download
\ No newline at end of file
+data_path=${1:-'.'}
+mkdir -p $data_path
+
+wget -P $data_path https://cernbox.cern.ch/index.php/s/SAHcpBDTp4P9jfO/download
+tar -xzf $data_path/download
```
Comment (V3RGANz) [inline get_csv.sh:5]: if you modify this script you should probably add changes to README.md

File: train.py [hunk 2]
```diff
@@ -2,84 +2,135 @@
...
+        D_real = self.discriminator(X_batch_real.float())
 
-            X_noise = torch.randn((batch_size, parameters.gan_noise_size), device=device)
-            X_batch_fake = generator(X_noise).detach()
+        epsilon = torch.rand(X_batch_real.shape[0], 1).expand(X_batch_real.size()).to(self.device)
 
-            predict = discriminator(X_batch_real.float())
-            loss = criterion(predict, y_real)
+        G_interpolation = epsilon * X_batch_real.float() + (1 - epsilon) * G_output.float()
+        G_interpolation = torch.autograd.Variable(G_interpolation, requires_grad=True)
```
Comment (V3RGANz) [inline train.py]: This is redundant. Deprecated in torch scince 0.4.0 https://pytorch.org/blog/pytorch-0_4_0-migration-guide/

File: train.py [hunk 3]
```diff
@@ -2,84 +2,135 @@
...
+        return D_loss
+
+    def generator_loss(self, noise_batch_shape):
+        X_noise = torch.randn(noise_batch_shape).to(self.device)
+        G_output = self.generator(X_noise)
+        D_output = self.discriminator(G_output.float())
+
+        G_loss = -torch.mean(torch.log(D_output + 1e-8))
+
+        return G_loss
```
Comment (V3RGANz) [inline train.py:95]: For vanilla GAN we have implemented functions or classes in torch for BCELoss calculations, that optimized and works well, what reason for implementing it manually?.

File: train.py [hunk 4]
```diff
@@ -2,84 +2,135 @@
...
-            loss.backward()
-            optimizer_d.step()
+        D_loss = torch.mean(D_fake) - torch.mean(D_real) + grad_penalty
 
-            metric_accum.disc_accum(loss.item(), torch.round(predict), y_fake)
+        return D_loss
 
-            if scheduler_d is not None:
-                scheduler_d.step()
+    def generator_loss(self, noise_size):
```
Comment (V3RGANz) [inline train.py:131]: the same as above, this loss function already implemented in torch, isn't it?

**Review by V3RGANz — CHANGES_REQUESTED** [review]
Almost OK, have you tested both vanilla GAN and WGAN-GP trained correctly after modifications?

**Comment by V3RGANz** [comment]
Also add instructions for running WGAN-GP to README.md, in case it runs in a different way

File: get_csv.sh (hunk 1 above)
Comment (eiofmv) [reply get_csv.sh:5]: No changes are needed to reproduce experiments described in paper.

File: train.py (hunk 3 above)
Comment (eiofmv) [reply train.py:95]: To provide the same interface for all classes.

File: train.py (hunk 4 above)
Comment (eiofmv) [reply train.py:131]: No.

File: train.py (hunk 2 above)
Comment (eiofmv) [reply train.py]: Fixed.

File: train.py (hunk 3 above)
Comment (V3RGANz) [reply train.py:95]: I meant something like
```
from torch.nn.functional import binary_cross_entropy

def generator_loss(self, noise_batch_shape):
    X_noise = torch.randn(noise_batch_shape, device=device)
    G_output = self.generator(X_noise)
    D_output = self.discriminator(G_output.float())
    target = torch.ones_like(D_output, device=device)

    return binary_cross_entropy(D_output, target)
```
this more obvious when you reading code
But I agree this is not important in this case actually, so do not change it if you think this is better now.

**Review by V3RGANz — APPROVED** [review]

**Review by ngushchin — APPROVED** [review]
